### ISSUE TITLE
package Prv: X and Obs: X should be picked first by provides

### ISSUE THREAD

**Issue by jsilhan** [openSUSE/libsolv#137]
Hi,
I would expect a package that provides X and obsoletes X to by picked first when searching by provide (when all provides are unversioned). So the preference in the testcase below would be `D (obsoletes) > B (provide == name) > A / C`. I can do a workaround using a *favor* but it would be great if the depsolver during resolution steps always behaved that way and didn't prefer this package in different rule.
```
repo system 0 testtags <inline>
repo available 0 testtags <inline>
#>=Pkg: A 1 1 noarch
#>=Prv: B
#>=Pkg: B 1 1 noarch
#>=Pkg: C 1 1 noarch
#>=Prv: B
#>=Pkg: D 1 1 noarch
#>=Obs: B
#>=Prv: B
system x86_64 rpm system
poolflags implicitobsoleteusescolors
solverflags allowvendorchange keepexplicitobsoletes bestobeypolicy keeporphans yumobsoletes
job install provides B
result transaction,problems <inline>
#>install A-1-1.noarch@available
```

**Comment by mlschroe** [comment]
Dunno. A also provides B. Why would it matter that a package named B has been obsoleted by D? Remember that obsoletes work on package names, not provides.

Now if D obsoleted A as well, the solver would ignore A and choose between C and D.

**Comment by jsilhan** [comment]
ok, in this case it's resolved fine so in the next `job update all packages` nothing gets updated. What about prioritizing the `package name == provide` to be picked by provide?
```
repo system 0 testtags <inline>
repo available 0 testtags <inline>
#>=Pkg: A 1 1 noarch
#>=Prv: B
#>=Pkg: B 1 1 noarch
system x86_64 rpm system
poolflags implicitobsoleteusescolors
solverflags allowvendorchange keepexplicitobsoletes bestobeypolicy keeporphans yumobsoletes
job install provides B
result transaction,problems <inline>
#>install B-1-1.noarch@available
```
currently `A` is installed.

**Comment by mlschroe** [comment]
Hmm, so that versioned provides are preferred? Why are they better?

Note that the solver does pick B if A provides "B = 1-0" instead of just "B".

**Comment by mlschroe** [comment]
Is this about rhbz#1096506? In that case you really should check for provides/obsoletes and not ask the solver for a provides. Just preferring the right package is not enough, you need to make sure that only packages with a correct obsoletes are taken into account.

So you need to check for provides/obsoletes and use SOLVER_SOLVABLE_ONE_OF if you find extra packages. And probably only do that if the user did not specify a specific version.

**Comment by jsilhan** [comment]
I've tried to prioritize the packages using `favor` [here](https://github.com/jsilhan/libhif/commit/e86977c485ba267ad6bd3a5a4a14b9a567c5d0e3). Is it correct? Do you have any example how to use `SOLVER_SOLVABLE_ONE_OF`? Is it something like `or` rich dep in spec? And I've found another issue https://github.com/openSUSE/libsolv/issues/139

**Comment by mlschroe** [comment]
About SOLVER_SOLVABLE_ONE_OF: solver jobs are *always* set of packages. There just exist multiple ways to tell the solver the desired package sets. This is just for convenience, it it easier to use SOLVER_SOLVABLE_NAME instead of calculating a package set with packages matching the name and then giving the solver the list.

SOLVER_SOLVABLE_ONE_OF works like this: you generate a list of package ids, than use pool_queuetowhatprovides to "internalize" the list into an id. You create a job of type SOLVER_SOLVABLE_ONE_OF with that id.

**Comment by jsilhan** [comment]
oh I see, I knew it's possible to install one package of given set but didn't know how to add ids into the job queue. Thanks for explanation. using disfavor in rhbz#1096506 works after #139 fix but we will use `SOLVER_SOLVABLE_ONE_OF` for installing packages based on given advisory.
